File: messaging/kafka/go/ops/producer.go
package main

import (
	"context"
	"fmt"
	"log"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/twmb/franz-go/pkg/kgo"
)

// pad — тот же паттерн, что в ../storage/producer.go: значение фиксированного
// размера (нужно для честного throughput/сжатия — реальный размер записи
// управляем мы, а не длина случайного мусора).
func pad(base string, targetBytes int, filler byte) string {
	if len(base) >= targetBytes {
		return base
	}
	return base + strings.Repeat(string(filler), targetBytes-len(base))
}

// newSeedProducer — простой acks=all клиент без явной тонкой настройки
// batch/linger/compression (эти параметры — предмет tuning-producer, а не
// seed/seed-continuous, которые лишь готовят данные для lag/tuning-consumer
// демонстраций).
func newSeedProducer(seeds []string) *kgo.Client {
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(seeds...),
		kgo.RequiredAcks(kgo.AllISRAcks()),
	)
	if err != nil {
		log.Fatalf("kgo.NewClient (seed producer): %v", err)
	}
	return cl
}

// seedFast — n записей асинхронно (Produce + один Flush в конце), максимально
// быстро. Используется, чтобы заранее наполнить топик данными для
// tuning-consumer (там тюнингуется КОНСЬЮМЕР, а не продюсер, поэтому сам
// сид должен быть быстрым и не влиять на измерения).
func seedFast(cl *kgo.Client, topic string, n int, valueBytes int, prefix string) {
	var acked, failed atomic.Int64
	start := time.Now()
	for i := 0; i < n; i++ {
		value := pad(fmt.Sprintf("%s-%08d-", prefix, i), valueBytes, 'v')
		rec := &kgo.Record{Topic: topic, Value: []byte(value)}
		cl.Produce(context.Background(), rec, func(_ *kgo.Record, err error) {
			if err != nil {
				failed.Add(1)
			} else {
				acked.Add(1)
			}
		})
	}
	if err := cl.Flush(context.Background()); err != nil {
		log.Fatalf("seedFast: Flush: %v", err)
	}
	elapsed := time.Since(start)
	fmt.Printf("[seed] topic=%s: отправлено %d записей (~%d байт значение) acked=%d failed=%d за %s\n",
		topic, n, valueBytes, acked.Load(), failed.Load(), elapsed)
}

// seedContinuous — фоновый продюсер для lag-демо: шлёт записи с целевым
// темпом ratePerSec в течение duration, затем останавливается. Печатает
// прогресс каждые ~2с, чтобы оркестрирующий bash-скрипт видел живой процесс
// и мог синхронизировать запуск lag-consume/поллинг kafka-consumer-groups.sh
// с фактическим стартом продюсинга, а не вслепую по sleep.
func seedContinuous(cl *kgo.Client, topic string, duration time.Duration, ratePerSec int, valueBytes int) {
	if ratePerSec <= 0 {
		ratePerSec = 1
	}
	interval := time.Second / time.Duration(ratePerSec)
	var sent atomic.Int64
	deadline := time.Now().Add(duration)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	progress := time.NewTicker(2 * time.Second)
	defer progress.Stop()
	i := 0
	fmt.Printf("[seed-continuous] СТАРТ topic=%s rate=%d/s duration=%s (~%d записей ожидается)\n", topic, ratePerSec, duration, int(duration.Seconds())*ratePerSec)
	for time.Now().Before(deadline) {
		select {
		case <-ticker.C:
			value := pad(fmt.Sprintf("lag-%08d-", i), valueBytes, 'l')
			i++
			rec := &kgo.Record{Topic: topic, Value: []byte(value)}
			cl.Produce(context.Background(), rec, func(_ *kgo.Record, err error) {
				if err == nil {
					sent.Add(1)
				} else {
					log.Printf("[seed-continuous] produce error: %v", err)
				}
			})
		case <-progress.C:
			fmt.Printf("[seed-continuous] прогресс: отправлено=%d\n", sent.Load())
		}
	}
	if err := cl.Flush(context.Background()); err != nil {
		log.Printf("[seed-continuous] Flush: %v", err)
	}
	fmt.Printf("[seed-continuous] ЗАВЕРШЕНО: всего отправлено %d записей за %s (целевой темп %d/s)\n", sent.Load(), duration, ratePerSec)
}

// newTuningProducer — клиент с явно управляемыми batch.size (ProducerBatchMaxBytes)
// / linger.ms (ProducerLinger) / compression.type (ProducerBatchCompression) —
// это и есть предмет демонстрации данного сценария, поэтому все три параметра
// приходят снаружи, а не из дефолтов библиотеки.
func newTuningProducer(seeds []string, batchBytes int32, linger time.Duration, compression kgo.CompressionCodec, clientID string) *kgo.Client {
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(seeds...),
		kgo.RequiredAcks(kgo.AllISRAcks()),
		kgo.ProducerBatchMaxBytes(batchBytes),
		kgo.ProducerLinger(linger),
		kgo.ProducerBatchCompression(compression),
		kgo.ClientID(clientID),
	)
	if err != nil {
		log.Fatalf("kgo.NewClient (tuning producer): %v", err)
	}
	return cl
}

// runTuningProducer — n записей АСИНХРОННО (не ProduceSync!) — синхронная
// отправка "запись за записью" вырождает batch.size/linger.ms в бессмысленные
// параметры (каждый батч был бы размера 1, см. тот же довод в
// ../storage/producer.go про produceBatchedAsync и компрессию). Здесь
// нарочно то же самое: чтобы batch/linger реально повлияли на throughput,
// клиенту нужно дать накопить настоящие батчи.
func runTuningProducer(cl *kgo.Client, topic string, n int, valueBytes int, label string) {
	var acked, failed atomic.Int64
	var wg sync.WaitGroup
	start := time.Now()
	for i := 0; i < n; i++ {
		value := pad(fmt.Sprintf("tune-%08d-", i), valueBytes, 't')
		rec := &kgo.Record{Topic: topic, Value: []byte(value)}
		wg.Add(1)
		cl.Produce(context.Background(), rec, func(_ *kgo.Record, err error) {
			defer wg.Done()
			if err == nil {
				acked.Add(1)
			} else {
				failed.Add(1)
			}
		})
	}
	wg.Wait()
	elapsed := time.Since(start)
	throughput := float64(n) / elapsed.Seconds()
	mbps := throughput * float64(valueBytes) / (1024 * 1024)
	fmt.Printf("[tuning-producer] %s n=%d value-bytes=%d -> acked=%d/%d failed=%d elapsed=%s throughput=%.1f msg/s (%.2f MB/s)\n",
		label, n, valueBytes, acked.Load(), n, failed.Load(), elapsed, throughput, mbps)
}

// quotaProduce — продюсит НЕПРЕРЫВНО (без искусственного rate limit) в течение
// duration под заданным client-id — именно client-id используется Kafka для
// сопоставления с quota-конфигом (entity-type=clients, entity-name=<client-id>).
// Измеряет ДОСТИГНУТЫЙ throughput за окно — до применения quota он упирается
// в реальную пропускную способность (сеть/диск), после применения — в
// producer_byte_rate, который выставляет ops-скрипт МЕЖДУ двумя прогонами.
func quotaProduce(seeds []string, topic, clientID string, duration time.Duration, valueBytes int) {
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(seeds...),
		kgo.RequiredAcks(kgo.AllISRAcks()),
		kgo.ClientID(clientID),
	)
	if err != nil {
		log.Fatalf("kgo.NewClient (quota producer): %v", err)
	}
	defer cl.Close()

	var acked, failed atomic.Int64
	var wg sync.WaitGroup
	start := time.Now()
	deadline := start.Add(duration)
	i := 0
	for time.Now().Before(deadline) {
		value := pad(fmt.Sprintf("quota-%08d-", i), valueBytes, 'q')
		i++
		rec := &kgo.Record{Topic: topic, Value: []byte(value)}
		wg.Add(1)
		cl.Produce(context.Background(), rec, func(_ *kgo.Record, err error) {
			defer wg.Done()
			if err == nil {
				acked.Add(1)
			} else {
				failed.Add(1)
			}
		})
	}
	wg.Wait()
	elapsed := time.Since(start)
	throughput := float64(acked.Load()) / elapsed.Seconds()
	mbps := throughput * float64(valueBytes) / (1024 * 1024)
	fmt.Printf("[quota-produce] client-id=%s duration=%s -> acked=%d failed=%d throughput=%.1f msg/s (%.2f MB/s)\n",
		clientID, elapsed, acked.Load(), failed.Load(), throughput, mbps)
}

func codecFromString(s string) kgo.CompressionCodec {
	switch s {
	case "none":
		return kgo.NoCompression()
	case "gzip":
		return kgo.GzipCompression()
	case "lz4":
		return kgo.Lz4Compression()
	case "zstd":
		return kgo.ZstdCompression()
	case "snappy":
		return kgo.SnappyCompression()
	default:
		log.Fatalf("неизвестный -compression=%q (none|gzip|lz4|zstd|snappy)", s)
		return kgo.NoCompression()
	}
}

File: messaging/kafka/java/ops/src/main/java/tech/khorost/kafka/ops/Producer.java
package tech.khorost.kafka.ops;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Продюсерские сценарии стенда #6 — зеркало ../../go/ops/producer.go: seed
 * (быстрое наполнение топика для lag/tuning-consumer), seed-continuous
 * (фоновый продюсер для lag-демо), tuning-producer (управляемые
 * batch.size/linger.ms/compression.type), quota-produce (непрерывная отправка
 * под заданным client.id для демонстрации quotas).
 */
public final class Producer {
    private Producer() {
    }

    static String pad(String base, int targetBytes, char filler) {
        if (base.length() >= targetBytes) return base;
        StringBuilder sb = new StringBuilder(base);
        while (sb.length() < targetBytes) sb.append(filler);
        return sb.toString();
    }

    private static Properties baseProducerProps(String clientId) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, Kafka.BOOTSTRAP);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        if (clientId != null && !clientId.isEmpty()) {
            props.put(ProducerConfig.CLIENT_ID_CONFIG, clientId);
        }
        return props;
    }

    public static void seedFast(String topic, int n, int valueBytes, String prefix) {
        Properties props = baseProducerProps(null);
        AtomicLong acked = new AtomicLong();
        AtomicLong failed = new AtomicLong();
        long start = System.nanoTime();
        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            for (int i = 0; i < n; i++) {
                String value = pad(String.format("%s-%08d-", prefix, i), valueBytes, 'v');
                producer.send(new ProducerRecord<>(topic, value), (meta, ex) -> {
                    if (ex == null) acked.incrementAndGet();
                    else failed.incrementAndGet();
                });
            }
            producer.flush();
        }
        double elapsedS = (System.nanoTime() - start) / 1e9;
        System.out.printf("[seed] topic=%s: отправлено %d записей (~%d байт значение) acked=%d failed=%d за %.3fs%n",
                topic, n, valueBytes, acked.get(), failed.get(), elapsedS);
    }

    public static void seedContinuous(String topic, long durationMs, int ratePerSec, int valueBytes) {
        Properties props = baseProducerProps(null);
        long intervalMs = Math.max(1, 1000 / Math.max(1, ratePerSec));
        AtomicLong sent = new AtomicLong();
        long deadline = System.currentTimeMillis() + durationMs;
        long lastLog = System.currentTimeMillis();
        int i = 0;
        System.out.printf("[seed-continuous] СТАРТ topic=%s rate=%d/s duration=%dms (~%d записей ожидается)%n",
                topic, ratePerSec, durationMs, (durationMs / 1000) * ratePerSec);
        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            while (System.currentTimeMillis() < deadline) {
                String value = pad(String.format("lag-%08d-", i), valueBytes, 'l');
                i++;
                producer.send(new ProducerRecord<>(topic, value), (meta, ex) -> {
                    if (ex == null) sent.incrementAndGet();
                });
                if (System.currentTimeMillis() - lastLog > 2000) {
                    System.out.printf("[seed-continuous] прогресс: отправлено=%d%n", sent.get());
                    lastLog = System.currentTimeMillis();
                }
                Kafka.sleep(intervalMs);
            }
            producer.flush();
        }
        System.out.printf("[seed-continuous] ЗАВЕРШЕНО: всего отправлено %d записей за %dms (целевой темп %d/s)%n",
                sent.get(), durationMs, ratePerSec);
    }

    public static void runTuningProducer(String topic, int n, int valueBytes, int batchBytes, int lingerMs, String compression, String clientId, String label) {
        Properties props = baseProducerProps(clientId);
        props.put(ProducerConfig.BATCH_SIZE_CONFIG, batchBytes);
        props.put(ProducerConfig.LINGER_MS_CONFIG, lingerMs);
        props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, compression);

        AtomicLong acked = new AtomicLong();
        AtomicLong failed = new AtomicLong();
        CountDownLatch latch = new CountDownLatch(n);
        long start = System.nanoTime();
        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            for (int i = 0; i < n; i++) {
                String value = pad(String.format("tune-%08d-", i), valueBytes, 't');
                producer.send(new ProducerRecord<>(topic, value), (meta, ex) -> {
                    if (ex == null) acked.incrementAndGet();
                    else failed.incrementAndGet();
                    latch.countDown();
                });
            }
            producer.flush();
            await(latch);
        }
        double elapsedS = (System.nanoTime() - start) / 1e9;
        double throughput = n / elapsedS;
        double mbps = throughput * valueBytes / (1024.0 * 1024.0);
        System.out.printf("[tuning-producer] %s n=%d value-bytes=%d -> acked=%d/%d failed=%d elapsed=%.3fs throughput=%.1f msg/s (%.2f MB/s)%n",
                label, n, valueBytes, acked.get(), n, failed.get(), elapsedS, throughput, mbps);
    }

    public static void quotaProduce(String topic, String clientId, long durationMs, int valueBytes) {
        Properties props = baseProducerProps(clientId);
        AtomicLong acked = new AtomicLong();
        AtomicLong failed = new AtomicLong();
        AtomicLong sentCount = new AtomicLong();
        long start = System.currentTimeMillis();
        long deadline = start + durationMs;
        try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {
            while (System.currentTimeMillis() < deadline) {
                long i = sentCount.getAndIncrement();
                String value = pad(String.format("quota-%08d-", i), valueBytes, 'q');
                producer.send(new ProducerRecord<>(topic, value), (meta, ex) -> {
                    if (ex == null) acked.incrementAndGet();
                    else failed.incrementAndGet();
                });
            }
            producer.flush();
        }
        double elapsedS = (System.currentTimeMillis() - start) / 1000.0;
        double throughput = acked.get() / elapsedS;
        double mbps = throughput * valueBytes / (1024.0 * 1024.0);
        System.out.printf("[quota-produce] client-id=%s duration=%.3fs -> acked=%d failed=%d throughput=%.1f msg/s (%.2f MB/s)%n",
                clientId, elapsedS, acked.get(), failed.get(), throughput, mbps);
    }

    private static void await(CountDownLatch latch) {
        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
